code/modules/ai/ai_behaviors/carbon/xeno/xeno.dm

```diff
@@ -0,0 +1,141 @@
+	for(var/x in GLOB.alive_xeno_list)
+		if(!xeno_parent.issamexenohive(x)) //Xenomorphs not in our hive will attack as well!
+			return_result += x
+	for(var/turret in GLOB.marine_turrets)
+		var/atom/atom_turret = turret
+		if(QDELETED(atom_turret))
+			continue
+		if(!get_dist(mob_parent, atom_turret) <= 8)
+			continue
+		if(mob_parent.z != atom_turret.z)
+			continue
+		return_result += turret
+	return return_result
+
+/datum/ai_behavior/carbon/xeno/do_process()
+	if(isainode(atom_to_walk_to) && length(get_targets()))
+		return REASON_TARGET_SPOTTED
+	if(istype(atom_to_walk_to, /mob/living/carbon/human) || istype(atom_to_walk_to, /obj/machinery/marine_turret))
+		return REASON_REFRESH_TARGET //We'll repick our targets as there could be more better targets to attack
+	return ..()
+
+/datum/ai_behavior/carbon/xeno/deal_with_obstacle()
+	if(!world.time > mob_parent.next_move)
+		return
+
+	var/list/things_nearby = range(mob_parent, 1) //Rather than doing multiple range() checks we can just archive it here for just this deal_with_obstacle
+	for(var/obj/structure/obstacle in things_nearby)
+		if(!(obstacle.resistance_flags & XENO_DAMAGEABLE))
+			var/mob/living/carbon/xenomorph/xeno = mob_parent
+			obstacle.attack_alien(xeno)
+			mob_parent.face_atom(obstacle)
+			xeno.changeNext_move(xeno.xeno_caste.attack_delay)
+			return
+
+	//Cheat mode: insta open airlocks
+	for(var/obj/machinery/door/airlock/lock in things_nearby)
+		if(!lock.density) //Airlock is already open no need to force it open again
+			continue
+		if(lock.operating) //Airlock already doing something
+			continue
+		if(lock.welded) //It's welded, can't force that open
+			continue
+		lock.open()
+		return //Don't try going on window frames after opening up airlocks dammit
+
+	//Teleport onto those window frames, we also can't attempt to attack said window frames so this isn't in the obstacles loop
+	for(var/obj/structure/window_frame/frame in things_nearby)
+		mob_parent.loc = frame.loc
+		return
+
+/datum/ai_behavior/carbon/xeno/attack_target()
+	if(world.time < mob_parent.next_move)
+		return
+	if(get_dist(atom_to_walk_to, mob_parent) > attack_range)
+		return
+	var/mob/living/carbon/xenomorph/xeno = mob_parent
+	mob_parent.face_atom(atom_to_walk_to)
+	if(ismob(atom_to_walk_to))
+		atom_to_walk_to.attack_alien(xeno)
+	else if(ismachinery(atom_to_walk_to))
+		var/obj/machinery/thing = atom_to_walk_to
+		if(!(thing.resistance_flags & XENO_DAMAGEABLE))
+			stack_trace("A xenomorph tried to attack a [atom_to_walk_to.name] that isn't considered XENO_DAMAGABLE according to resistance flags.")
+		thing.attack_alien(xeno)
+	xeno.changeNext_move(xeno.xeno_caste.attack_delay)
+
+/datum/ai_behavior/carbon/xeno/get_new_state(reason_for)
+	switch(reason_for) //Lets look for targets when we're recently done doing an action, gotta find enemies to kill
+		if(REASON_FINISHED_NODE_MOVE, REASON_TARGET_KILLED, REASON_TARGET_SPOTTED, REASON_REFRESH_TARGET)
```

Comment:
Ok, what is the idea for this check? It's checking all the possible values.
Did you want something like `if(!reason_for) return ..()` or do you expect for there to be more reasons in the future that would require skipping and being processed by the parent instead?